Add `mypy` static typing

More info: http://mypy-lang.org/

Config:
```
-   repo: https://github.com/pre-commit/mirrors-mypy
    rev: v0.910
    hooks:
    -   id: mypy
```

Current error log:
```
src/server/tests/utils.py:8: error: Incompatible types in assignment (expression has type "Callable[[Dict[_KT, _VT], _KT, _VT], None]", base class "object" defined the type as "Callable[[object, str, Any], None]")
src/server/tests/utils.py:9: error: Incompatible types in assignment (expression has type "Callable[[Dict[_KT, _VT], _KT], None]", base class "object" defined the type as "Callable[[object, str], None]")
src/server/backend/helpers/command_runner.py:83: error: Incompatible return value type (got "Optional[str]", expected "str")
src/server/backend/helpers/mocks/pywifi_mock.py:54: error: Need type annotation for "results" (hint: "results: List[<type>] = ...")
src/server/backend/helpers/finalise.py:24: error: Incompatible return value type (got "str", expected "int")
src/server/backend/helpers/finalise.py:29: error: Incompatible types in assignment (expression has type "None", variable has type "str")
src/server/backend/helpers/finalise.py:32: error: Incompatible types in assignment (expression has type "None", variable has type "str")
src/server/backend/helpers/finalise.py:35: error: Incompatible return value type (got "str", expected "int")
src/server/backend/helpers/language.py:14: error: Need type annotation for "utf8_locales" (hint: "utf8_locales: List[<type>] = ...")
src/server/backend/helpers/language.py:34: error: Incompatible return value type (got "None", expected "str")
src/server/backend/helpers/language.py:47: error: Need type annotation for "utf8_locales" (hint: "utf8_locales: List[<type>] = ...")
src/server/backend/helpers/about.py:50: error: Need type annotation for "most_unstable_source" (hint: "most_unstable_source: Dict[<type>, <type>] = ...")
src/server/backend/helpers/about.py:58: error: Incompatible return value type (got "Optional[Any]", expected "str")
src/server/backend/helpers/wifi_country.py:30: error: Incompatible types in assignment (expression has type "None", variable has type "str")
src/server/backend/helpers/timezone.py:41: error: Incompatible return value type (got "Optional[str]", expected "str")
src/server/backend/helpers/keyboard.py:29: error: Incompatible return value type (got "Tuple[Optional[str], Optional[str]]", expected "Tuple[str, str]")
src/server/backend/helpers/device_registration.py:9: error: Library stubs not installed for "requests" (or incompatible with Python 3.9)
src/server/backend/helpers/device_registration.py:9: note: Hint: "python3 -m pip install types-requests"
src/server/backend/helpers/device_registration.py:9: note: (or run "mypy --install-types" to install all missing stub packages)
src/server/backend/helpers/device_registration.py:9: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
src/server/backend/helpers/wifi_manager.py:40: error: Missing return statement
src/server/backend/helpers/wifi_manager.py:89: error: Incompatible types in assignment (expression has type "float", variable has type "int")
src/server/backend/helpers/os_updater.py:13: error: Name "apt.progress.base.AcquireProgress" is not defined
src/server/backend/helpers/os_updater.py:35: error: Name "apt.progress.base.InstallProgress" is not defined
src/server/backend/helpers/os_updater.py:66: error: "None" has no attribute "update"
src/server/backend/helpers/os_updater.py:67: error: "None" has no attribute "open"
src/server/backend/helpers/os_updater.py:82: error: "None" has no attribute "upgrade"
src/server/backend/helpers/os_updater.py:83: error: "None" has no attribute "upgrade"
src/server/backend/helpers/os_updater.py:85: error: "None" has no attribute "install_count"
src/server/backend/helpers/os_updater.py:87: error: "None" has no attribute "required_download"
src/server/backend/helpers/os_updater.py:90: error: "None" has no attribute "required_space"
src/server/backend/routes.py:250: error: Item "None" of "Optional[Any]" has no attribute "route"
Found 29 errors in 13 files (checked 51 source files)
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -18,7 +18,7 @@ repos:
       args: ["--profile", "black"]
 
 -   repo: https://github.com/ambv/black
-    rev: 21.6b0
+    rev: 21.7b0
     hooks:
     - id: black
 

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -39,3 +39,5 @@ repos:
     rev: v0.910
     hooks:
     -   id: mypy
+        additional_dependencies: [types-all]
+        args: ["--python-version=3.7", "--ignore-missing-imports", "--install-types", "--non-interactive"]

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -34,3 +34,8 @@ repos:
     - id: flake8
       name: flake8
       args: ["--max-line-length=150"]
+
+-   repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v0.910
+    hooks:
+    -   id: mypy

--- a/src/server/backend/__init__.py
+++ b/src/server/backend/__init__.py
@@ -2,7 +2,7 @@ from flask import Flask
 from flask_cors import CORS
 from flask_sockets import Sockets
 
-sockets = None
+sockets: Sockets
 
 
 def create_app(test=False):

--- a/src/server/backend/helpers/about.py
+++ b/src/server/backend/helpers/about.py
@@ -47,7 +47,7 @@ def device_update_channel() -> str:
 
         return {}
 
-    most_unstable_source = {}
+    most_unstable_source: dict[str, str] = {}
     for line in _get_file_lines("/etc/apt/sources.list.d/pi-top.list"):
         source_data = get_source_from_line(line)
         if source_data and source_data.get("priority") >= most_unstable_source.get(
@@ -55,7 +55,7 @@ def device_update_channel() -> str:
         ):
             most_unstable_source = source_data
 
-    return most_unstable_source.get("name")
+    return most_unstable_source.get("name", "")
 
 
 def device_serial_number():

--- a/src/server/backend/helpers/about.py
+++ b/src/server/backend/helpers/about.py
@@ -1,5 +1,6 @@
 from json import load as jload
 from re import match
+from typing import Dict
 
 from pitopcommon.common_names import DeviceName
 
@@ -47,7 +48,7 @@ def device_update_channel() -> str:
 
         return {}
 
-    most_unstable_source: dict[str, str] = {}
+    most_unstable_source: Dict[str, str] = {}
     for line in _get_file_lines("/etc/apt/sources.list.d/pi-top.list"):
         source_data = get_source_from_line(line)
         if source_data and source_data.get("priority") >= most_unstable_source.get(

--- a/src/server/backend/helpers/command_runner.py
+++ b/src/server/backend/helpers/command_runner.py
@@ -25,7 +25,7 @@ def run_command(
         % (command_str, timeout, check, capture_output, lower_priority)
     )
 
-    resp_stdout = None
+    resp_stdout = ""
 
     env_plus_display = environ.copy()
     env_plus_display["DISPLAY"] = ":0"

--- a/src/server/backend/helpers/device_registration.py
+++ b/src/server/backend/helpers/device_registration.py
@@ -1,5 +1,3 @@
-#!/usr/bin/python3
-
 import json
 import os
 from pathlib import Path

--- a/src/server/backend/helpers/finalise.py
+++ b/src/server/backend/helpers/finalise.py
@@ -16,7 +16,7 @@ from .paths import (
 )
 
 
-def available_space() -> int:
+def available_space() -> str:
     PTLogger.info("Function: available_space()")
     out = run_command("df --block-size=1 --output=avail '/'", timeout=2).splitlines()
 
@@ -26,12 +26,9 @@ def available_space() -> int:
     if len(out) == 2:
         space = out[1].strip()
     else:
-        space = None
+        space = ""
 
-    if space == "":
-        space = None
-
-    PTLogger.info("Available Space: '%s'" % space)
+    PTLogger.info(f"Available Space: '{space}'")
     return space
 
 

--- a/src/server/backend/helpers/keyboard.py
+++ b/src/server/backend/helpers/keyboard.py
@@ -1,5 +1,5 @@
 from fileinput import input as finput
-from typing import Tuple
+from typing import Optional, Tuple
 
 from pitopcommon.logger import PTLogger
 
@@ -7,9 +7,9 @@ from .command_runner import run_command
 from .paths import default_keyboard_conf
 
 
-def current_keyboard_layout() -> Tuple[str, str]:
+def current_keyboard_layout() -> Tuple[str, Optional[str]]:
     PTLogger.info("Function: current_keyboard_layout()")
-    layout_code = None
+    layout_code = ""
     variant = None
 
     command = "setxkbmap -query"

--- a/src/server/backend/helpers/language.py
+++ b/src/server/backend/helpers/language.py
@@ -1,4 +1,5 @@
 from re import search
+from typing import List
 
 from pitopcommon.logger import PTLogger
 
@@ -11,7 +12,7 @@ def list_locales_supported() -> list:
 
     with open(supported_locales()) as file:
         lines = [line.rstrip() for line in file]
-        utf8_locales: list[str] = list()
+        utf8_locales: List[str] = list()
 
         for line in lines:
             # keep lines NOT starting with @ and containing .UTF-8
@@ -44,7 +45,7 @@ def list_locales_available() -> list:
 
     with open(locales_gen()) as file:
         lines = [line.rstrip() for line in file]
-        utf8_locales: list[str] = list()
+        utf8_locales: List[str] = list()
 
         for line in lines:
             # get just locale code from lines matching utf-8 locale regex

--- a/src/server/backend/helpers/language.py
+++ b/src/server/backend/helpers/language.py
@@ -11,7 +11,7 @@ def list_locales_supported() -> list:
 
     with open(supported_locales()) as file:
         lines = [line.rstrip() for line in file]
-        utf8_locales = list()
+        utf8_locales: list[str] = list()
 
         for line in lines:
             # keep lines NOT starting with @ and containing .UTF-8
@@ -31,7 +31,7 @@ def current_locale() -> str:
     with open(default_locale()) as file:
         lang_lines = [line.rstrip() for line in file if line.startswith("LANG=")]
         if len(lang_lines) == 0:
-            return None
+            return ""
 
         # Get last one, in case there are multiple
         locale = lang_lines[-1].split("=")[1].split(".UTF-8")[0]
@@ -44,7 +44,7 @@ def list_locales_available() -> list:
 
     with open(locales_gen()) as file:
         lines = [line.rstrip() for line in file]
-        utf8_locales = list()
+        utf8_locales: list[str] = list()
 
         for line in lines:
             # get just locale code from lines matching utf-8 locale regex

--- a/src/server/backend/helpers/mocks/pywifi_mock.py
+++ b/src/server/backend/helpers/mocks/pywifi_mock.py
@@ -51,7 +51,7 @@ class PyWiFiProfile:
 
 class PyWiFiInterfaceMock:
     state = PyWiFiConstant.IFACE_INACTIVE
-    results = []
+    results: list[PyWiFiProfile] = []
 
     def __init__(self):
         self._wifi_ctrl = PyWiFiUtil()

--- a/src/server/backend/helpers/mocks/pywifi_mock.py
+++ b/src/server/backend/helpers/mocks/pywifi_mock.py
@@ -1,4 +1,5 @@
 from enum import Enum
+from typing import List
 
 
 class PyWiFiConstant(Enum):
@@ -51,7 +52,7 @@ class PyWiFiProfile:
 
 class PyWiFiInterfaceMock:
     state = PyWiFiConstant.IFACE_INACTIVE
-    results: list[PyWiFiProfile] = []
+    results: List[PyWiFiProfile] = []
 
     def __init__(self):
         self._wifi_ctrl = PyWiFiUtil()

--- a/src/server/backend/helpers/os_updater.py
+++ b/src/server/backend/helpers/os_updater.py
@@ -10,7 +10,7 @@ from .system_clock import is_system_clock_synchronized, synchronize_system_clock
 (apt, apt.progress, apt_pkg) = get_apt()
 
 
-class FetchProgress(apt.progress.base.AcquireProgress):
+class FetchProgress(apt.progress.base.AcquireProgress):  # type: ignore
     def __init__(self, callback):
         apt.progress.base.AcquireProgress.__init__(self)
         self.callback = callback
@@ -32,7 +32,7 @@ class FetchProgress(apt.progress.base.AcquireProgress):
         return apt.progress.base.AcquireProgress.pulse(self, owner)
 
 
-class InstallProgress(apt.progress.base.InstallProgress):
+class InstallProgress(apt.progress.base.InstallProgress):  # type: ignore
     def __init__(self, callback):
         apt.progress.base.InstallProgress.__init__(self)
         self.callback = callback
@@ -49,7 +49,7 @@ class OSUpdater:
     lock = False
 
     def __init__(self) -> None:
-        self.cache = None
+        self.cache: apt.Cache  # type: ignore
         self.CONFIG_DIRECTORY = "/etc/pi-top/pt-os-updater/"
         self.LAST_CHECKED_CONFIG_FILE = self.CONFIG_DIRECTORY + "last_checked_date"
 

--- a/src/server/backend/helpers/timezone.py
+++ b/src/server/backend/helpers/timezone.py
@@ -27,7 +27,7 @@ def get_all_timezones() -> list:
 
 def get_current_timezone() -> str:
     PTLogger.info("Function: get_current_timezone()")
-    tz_string = None
+    tz_string = ""
 
     if use_test_path():
         return "Europe/London"

--- a/src/server/backend/helpers/wifi_country.py
+++ b/src/server/backend/helpers/wifi_country.py
@@ -26,9 +26,6 @@ def current_wifi_country() -> str:
     ).strip()
     PTLogger.info("Current Wi-Fi country: '%s'" % wifi_country)
 
-    if wifi_country == "":
-        wifi_country = None
-
     return wifi_country
 
 

--- a/src/server/backend/helpers/wifi_manager.py
+++ b/src/server/backend/helpers/wifi_manager.py
@@ -50,6 +50,7 @@ class WifiManager:
             return IfaceStatus.CONNECTING
         elif status_int == 4:
             return IfaceStatus.CONNECTED
+        return IfaceStatus.DISCONNECTED
 
     def is_inactive(self) -> bool:
         return self.get_status() in (IfaceStatus.DISCONNECTED, IfaceStatus.INACTIVE)
@@ -72,7 +73,7 @@ class WifiManager:
         silent: bool = False,
     ) -> None:
         sleep_time = 0.1
-        time_waited = 0
+        time_waited = 0.0
 
         while not condition_func() if condition_true else condition_func():
             text = "Waiting"

--- a/src/server/backend/routes.py
+++ b/src/server/backend/routes.py
@@ -73,7 +73,7 @@ from .helpers.wifi_manager import (
 
 
 def abort_on_no_data(data):
-    if data is None:
+    if data is None or (isinstance(data, str) and len(data) == 0):
         abort(400)
     return jdumps(data)
 

